detect_set_scale: use ratio of drawable size and window size for auto…

edwintorok opened the pull request:
…-detection

On Fedora37, wayland with desktop scale set to 200% on a 4K screen SDL2 would detect a DPI of 92.

https://wiki.libsdl.org/SDL2/SDL_GetDisplayDPI says that the function is unreliable, and that the ratio of SDL_GL_GetDrawableSize and SDL_GetWindowSize should be used instead.

There is already a dpi_xscale/dpi_yscale that is set, but setting DPI at that point would be too late: the layout and fonts would already be loaded with the old scale.

The only way to get the desired scale was to use BOGUE_SCALE=2 env var. (Calling the various resize/set_size functions doesn't correctly refresh everything with the new scale, apparently has to be done at initialization time).

So create a fake 10 pixel window to detect scale for now.

TODO (out of scope of this PR):
The initialization order should probably be adjusted to initialize anything that depends on window size after Draw.init. Also it'd be nice if the window could react to DPI changes when moved from one monitor to another.

edwintorok committed:
detect_set_scale: use ratio of drawable size and window size for auto-detection

On Fedora37, wayland with desktop scale set to 200% on a 4K screen
SDL2 would detect a DPI of 92.

https://wiki.libsdl.org/SDL2/SDL_GetDisplayDPI says that the function is
unreliable, and that the ratio of SDL_GL_GetDrawableSize and
SDL_GetWindowSize should be used instead.

There is already a dpi_xscale/dpi_yscale that is set, but setting DPI at
that point would be too late: the layout and fonts would already be
loaded with the old scale.

The only way to get the desired scale was to use BOGUE_SCALE=2 env var.
(Calling the various resize/set_size functions doesn't correctly refresh
everything with the new scale, apparently has to be done at
initialization time).

So create a fake 10 pixel window to detect scale for now.

For best cross platform compatibility use the ratio computed above as a
multiplier for the DPI obtained from the system.

This only works reliably with wayland: X11, Mac OS X and other platforms
already seem to report the correct DPI.

Signed-off-by: Edwin Török <edwin@etorok.net>

edwintorok commented:
Tested with 'boguex 44', now I get same window, text and button text size as with BOGUE_SCALE=2, while previously everything was half the size it was supposed to be.

sanette commented:
Thanks for the report and PR! we have to be careful here, because there is already a similar mechanism that was implemented in order to correct the "false DPI" reported by Mac OS retina screens, see the `Draw.init` function
https://github.com/sanette/bogue/blob/0ba3637108ea8355e05e86251fd45ebee38a1c7f/lib/b_draw.ml#L1248

* could you post the log of `BOGUE_DEBUG=true boguex 44`?
* could you try to replace there `Sdl.get_renderer_output_size` by `Sdl.gl_get_drawable_size`?

edwintorok commented:
Here is the log with only my patch: [log.txt](https://github.com/sanette/bogue/files/10613729/log.txt)
And here is the log with also this diff:
```diff
diff --git a/lib/b_draw.ml b/lib/b_draw.ml
index 6c24cf287..c0e280991 100644
--- a/lib/b_draw.ml
+++ b/lib/b_draw.ml
@@ -1258,7 +1258,7 @@ let init ?window ?(name="BOGUE Window") ?fill ?x ?y ~w ~h () =
       | Ok w -> printd debug_graphics "Using existing renderer"; w
       | Error _ ->
         go (Sdl.create_renderer ~flags:Sdl.Renderer.targettexture win) in
-  let rw, rh = go (Sdl.get_renderer_output_size renderer) in
+  let rw, rh = Sdl.gl_get_drawable_size win in
   if window = None && (rw, rh) <> (w,h) then begin
     dpi_xscale := float rw /. float w;
     dpi_yscale := float rh /. float h;
```
[log-gl_get_drawable.txt](https://github.com/sanette/bogue/files/10613734/log-gl_get_drawable.txt)

sanette commented:
thanks! so both logs show indeed that the (original) code detected the false DPI:
```
[56][0] :	 Graphics; Warning: This display imposes a hard scaling of (2.000000,2.000000).
```
so I'm a bit puzzled why the scale was not properly corrected. I'll try to investigate this when I get some time (and a Mac with retina screen so that we make sure we don't introduce a regression on that side)

sanette commented:
I checked this PR on my machine and it does not work correctly: everything is tiny. Indeed, the new dpi detection gives me a scale=1 (because on X11 the window size is correctly respected and corresponds to drawable size), while my normal DPI  is around 277, which should give a scale of about 2.5

edwintorok commented:
Thanks, looks like finding something cross platform here will be difficult, I'll set this PR back to draft for now.

sanette commented:
Actually I think that a better way to compute the DPI is to *multiply* the result of SDL_GetDisplayDPI by the ratio (SDL_GL_GetDrawableSize / SDL_GetWindowSize), could you try this? This should work for you and I, but maybe requires some further tweaks for MacOS

edwintorok commented:
I've implemented the multiplication and this indeed works better in the sense that the 'bogue' window stays the same size. However now that happens even if the system-wide scale is set to 100% (because then it will report a higher DPI).
Also the auto-computed scale will be fractional (1.91), so had to tweak INT_SCALE a bit to round that to nearest instead of always down.

(the fractional scale might actually be correct, but most things will work better if the scale is an integer multiple, or at least if font sizes end up being integers).

I also changed the default DPI to 96 (that seems to be almost hardcoded in Xorg these days), not sure why the default DPI was  110, is that the default on Mac OS?

I'm not sure whether there is a cross-platform way to detect the actual scale set by the desktop (X11 apps typically have to rely on environment variables for that, and those env vars are different for each desktop environment, and wayland may have some API here but apparently not exposed well enough by SDL2).

Do my changes work on Mac OS X for you?

sanette commented:
> not sure why the default DPI was 110,

this was simply out of experiment. With 96, the scale would be too big on my machine, leading to large fonts compared to the rest of my desktop.  

My understanding is that the reference of 96dpi was used long time ago and screen were not sharp, nowadays many people seem to prefer smaller fonts. Of course this is a personal interpretation, but this seems to be consistent with what my KDE desktop does. With dpi=277 and reference 96dpi, I should set a scale=2.9 which will give my huge fonts and windows. (especially with integer rounding 3). I much prefer scale=2.5 or even 2.4

For instance this [page](https://www.tenforums.com/tutorials/5990-change-dpi-scaling-level-displays-windows-10-a.html) says: "Almost all desktop displays and most current laptop displays are in the range of 95-110 DPI; for these devices, no scaling is required, and Windows sets a scale factor of 100%."

this is why I chose 110 as a limit, but of course it's difficult to fit all cases.

This is why also allowing fractional scaling is important. It's not a problem with fonts. Font size are always integers in Bogue, even when the scale is not, and this allows a better choice than when scale is an integer. The only problem I found with fractional scaling is games with tilings. Sometimes the tiles are not perfectly adjacent if the scale is not an integer.

sanette commented:
> I've implemented the multiplication and this indeed works better in the sense that the 'bogue' window stays the same size.

Great, thanks!
could you post a screenshot of boguex 29? (and check if all radio button react normally?)

sanette commented:
> had to tweak INT_SCALE a bit to round that to nearest instead of always down.

the reason why I chose floor instead of round is that when you open a desktop app, I find it better to have slightly too small font and see the whole window, rather than slightly too big font and the app window ends up being too large for your screen, so the app becomes quickly unusable. My "good" scale is 2.5 so if you round it to 3 this problem is quite likely to happen. It happened to me several times with gtk apps and it's unpleasant for the user experience. 

Of course this is debatable. I agree that with 1.9 you would like to have 2, the risk of the window being larger than the screen is not so big. Maybe we can decide that we round up to 2 only above 1.75, or something like this. (the threshold could also be a proportion of the scale)

edwintorok commented:
Could multiply auto-detected scale by 4, round, then divide by 4 again to give 0.25 increments for scale, and leave DPI to 110 as it was.
In the end it is difficult to try to know what the user expects and what they've set for the rest of their desktop.
I'll do a few more tests and attach some screenshots.

Good to know that font sizes are always integers in bogue, that removes most of my worries about fractional scaling (assuming the rest of the UI is then proportional to the font size).

edwintorok committed:
Draw.detect_set_scale: round to nearest 0.25 multiple for SCALE

INT_SCALE may be too coarse: either 1x, 2x or 3x, etc.
However sometimes 1.75 or 2.5 is a more appropriate scale.

INT_SCALE can still be used if you want to floor to nearest integer.

Signed-off-by: Edwin Török <edwin@etorok.net>

edwintorok commented:
Screenshots of various configurations on a 4K, 24" screen:

Wayland, 100% Gnome scale: ![bogue_29_wayland_100%](https://user-images.githubusercontent.com/721894/217963977-d1588d42-312d-47da-b0c5-e976e41e7964.png)

Wayland, 200% Gnome scale: ![bogue_29_wayland_200%](https://user-images.githubusercontent.com/721894/217963980-7bacdb66-18de-4251-b40e-875b69536c46.png)

Wayland, 300% Gnome scale: ![bogue_29_wayland_300%](https://user-images.githubusercontent.com/721894/217963983-f93ddf4b-56e7-4a6a-acaa-8440ab802aba.png)

Xorg, 100% Gnome scale: ![bogue_29_xorg_100%](https://user-images.githubusercontent.com/721894/217963984-c87b6c9f-202f-4567-a8cc-a2ad4abe12de.png)

Xorg, 200% Gnome scale: ![bogue_29_xorg_200%](https://user-images.githubusercontent.com/721894/217963987-76dc8d2e-dba4-4998-a90a-76a227fabb08.png)

The font size in the title changes (this is probably outside of bogue's control), however the window itself stays approximately the same physical size.
(note however that on 100% desktop scale on a 4K screen now the 'bogue' window is 2x scale and readable, while every other window is 1x and tiny, so although 'bogue' matches how I like to see things it explicitly doesn't match the user's choice of scale anymore, at least not by default, they can of course always use BOGUE_SCALE to override.
It seems to match's bogue intent on providing a UI in units of ~110 DPI pixels)

sanette commented:
So today I could borrow a MacBook Pro, and, well... as expected the bogex example are twice too big with this PR...

[146][0] :	 Graphics: SDL Video initialized
[147][0] :	 Graphics: DPI from system: 227
[167][0] :	 Graphics: Autodetect multiplier: window(10,10), drawable(20,20)
[167][0] :	 Graphics; Warning: Using SCALE=4.250000

Here the DPI=227 is correctly detected, so the multiplier should not be applied...
we have to think more for a Multiplatform solution!

sanette commented:
I guess we have to do it case by case. Could you tell me the name of your SDL video driver?
```
utop
#require "tsdl";;
open Tsdl;;
Sdl.(init Init.video);;
Sdl.get_current_video_driver ();;`
```

edwintorok commented:
Thanks for the testing. The default video driver is `wayland`, and I can also choose to login with Xorg in which case the SDL video driver is `x11`.

sanette commented:
the problem seems to be related to SDL Wayland scale support. Could you try this under Wayland
```
Sdl.(init Init.video);;
let (w,r) = Sdl.create_window_and_renderer ~w:20 ~h:20 Sdl.Window.(allow_highdpi + opengl) |> function | Ok c -> c | _ -> failwith "error";;
Sdl.render_get_scale r;;
```

sanette commented:
As a workaround, I suggest to multiply by the "autodetected multiplier" *only* if `Sdl.get_current_video_driver () = "wayland"`

edwintorok commented:
~Hmm, something must've changed. I now have SDL2 2.26.2 (upgraded from 2.26.0 recently), and now even without these changes the SCALE gets correctly detected as 1.68 (instead of 1 like previously).
It still works correctly with these changes too, but if they're causing problems on other platforms then maybe drop the multiplication idea and instead just do the rounding to nearest 0.25 scale?~ nevermind I was still in X11, while I thought I was in wayland (interestingly the wayland driver does work under X11 too..)

edwintorok commented:
I've implemented the suggestion to do this only when driver is wayland. Thanks for your patience on investigating this.
(I got confused previously because SDL_VIDEODRIVER=x11 only sets the video driver of screen 1 to x11, screen 0 stayed with the wayland driver...)

sanette commented:
thanks!

sanette commented:
Oh well..., I just installed the Wayland plasma desktop to test on my laptop, and when using SDL_VIDEODRIVER="wayland", Bogue windows become way too big
```
[53][0] :        Graphics: SDL Video initialized
[53][0] :        Graphics: DPI from system: 277
[113][0] :       Graphics: Autodetect multiplier: window(10,10), drawable(30,30)
[113][0] :       Graphics; Warning: Using SCALE=7.500000
```
Here the DPI=277 was correctly detected, hence the multiplier made the scale insanely large :(
I guess it's impossible to cover all cases. Especially on Wayland where things are not very stable.

edwintorok commented:
what OS, and SDL2 version are you using? Also what is the output of 'wayland-info' (in wayland-utils package for me).
Here is my output:
```
interface: 'wl_compositor',                              version:  5, name:  1
interface: 'wl_drm',                                     version:  2, name:  2
interface: 'wl_shm',                                     version:  1, name:  3
	formats: 'XB4H'(0x48344258) 'AB4H'(0x48344241) 'XR4H'(0x48345258) 'AR4H'(0x48345241) 'XB30'(0x30334258) 'AB30'(0x30334241) 'XR30'(0x30335258) 'AR30'(0x30335241) RGB565 'XB24'(0x34324258) 'AB24'(0x34324241) XRGB8888 ARGB8888
interface: 'wl_output',                                  version:  3, name:  4
	x: 3840, y: 0, scale: 2,
	physical_width: 530 mm, physical_height: 300 mm,
	make: 'DEL', model: 'DELL P2415Q',
	subpixel_orientation: unknown, output_transform: normal,
	mode:
		width: 3840 px, height: 2160 px, refresh: 59.996 Hz,
		flags: current preferred
interface: 'zxdg_output_manager_v1',                     version:  3, name:  6
	xdg_output_v1
		output: 27
		name: 'HDMI-1'
		description: 'LG Electronics 27"'
		logical_x: 0, logical_y: 0
		logical_width: 3840, logical_height: 2160
	xdg_output_v1
		output: 4
		name: 'DP-2'
		description: 'Dell Inc. 24"'
		logical_x: 3840, logical_y: 0
		logical_width: 3840, logical_height: 2160
interface: 'wl_data_device_manager',                     version:  3, name:  7
interface: 'zwp_primary_selection_device_manager_v1',    version:  1, name:  8
interface: 'wl_subcompositor',                           version:  1, name:  9
interface: 'xdg_wm_base',                                version:  4, name: 10
interface: 'gtk_shell1',                                 version:  5, name: 11
interface: 'wp_viewporter',                              version:  1, name: 12
interface: 'zwp_pointer_gestures_v1',                    version:  3, name: 13
interface: 'zwp_tablet_manager_v2',                      version:  1, name: 14
interface: 'wl_seat',                                    version:  8, name: 15
	name: seat0
	capabilities: pointer keyboard
	keyboard repeat rate: 33
	keyboard repeat delay: 500
interface: 'zwp_relative_pointer_manager_v1',            version:  1, name: 16
interface: 'zwp_pointer_constraints_v1',                 version:  1, name: 17
interface: 'zxdg_exporter_v1',                           version:  1, name: 18
interface: 'zxdg_importer_v1',                           version:  1, name: 19
interface: 'zwp_linux_dmabuf_v1',                        version:  4, name: 20
	formats:
	'XB4H'(0x48344258), modifier: 0x00ffffffffffffff
	'AB4H'(0x48344241), modifier: 0x00ffffffffffffff
	'RG16'(0x36314752), modifier: 0x00ffffffffffffff
	'XB30'(0x30334258), modifier: 0x00ffffffffffffff
	'XR30'(0x30335258), modifier: 0x00ffffffffffffff
	'AB30'(0x30334241), modifier: 0x00ffffffffffffff
	'AR30'(0x30335241), modifier: 0x00ffffffffffffff
	'XB24'(0x34324258), modifier: 0x00ffffffffffffff
	'XR24'(0x34325258), modifier: 0x00ffffffffffffff
	'AB24'(0x34324241), modifier: 0x00ffffffffffffff
	'AR24'(0x34325241), modifier: 0x00ffffffffffffff
interface: 'wp_single_pixel_buffer_manager_v1',          version:  1, name: 21
interface: 'zwp_keyboard_shortcuts_inhibit_manager_v1',  version:  1, name: 22
interface: 'zwp_text_input_manager_v3',                  version:  1, name: 23
interface: 'wp_presentation',                            version:  1, name: 24
	presentation clock id: 1 (CLOCK_MONOTONIC)
interface: 'xdg_activation_v1',                          version:  1, name: 25
interface: 'wl_output',                                  version:  3, name: 27
	x: 0, y: 0, scale: 2,
	physical_width: 600 mm, physical_height: 340 mm,
	make: 'GSM', model: 'LG Ultra HD',
	subpixel_orientation: unknown, output_transform: normal,
	mode:
		width: 3840 px, height: 2160 px, refresh: 60.000 Hz,
		flags: current preferred
```

I can try to see whether I can get something similar to yours on my system (for me the DPI in wayland changes based on user configured desktop scale, at least in Gnome, while for you it looks like wayland reports the true DPI).

[I'm thinking we could parse the output of wayland-info, if installed, when the driver is wayland, like we already do or 'xdpyinfo' and that'd avoid any SDL2 issues with what DPI it presents to us. Assuming the info from the monitor is correct, which in my case isn't entirely: the LG monitor reports itself as 27" where in reality it is 24" (bug in monitor's EDID apparently, there is a 27" model too), but the Dell reports it correctly.]

sanette commented:
I'm using KDE Neon

```
interface: 'wl_compositor',                              version:  5, name:  1
interface: 'zwp_tablet_manager_v2',                      version:  1, name:  2
interface: 'zwp_keyboard_shortcuts_inhibit_manager_v1',  version:  1, name:  3
interface: 'xdg_wm_base',                                version:  4, name:  5
interface: 'zwlr_layer_shell_v1',                        version:  3, name:  6
interface: 'zxdg_decoration_manager_v1',                 version:  1, name:  7
interface: 'wp_viewporter',                              version:  1, name:  8
interface: 'wl_shm',                                     version:  1, name:  9
        formats: 'XB48'(0x38344258) 'AB48'(0x38344241) 'XB30'(0x30334258) 'AB30'(0x30334241) 'XR30'(0x30335258) 'AR30'(0x30335241) XRGB8888 ARGB8888
interface: 'wl_seat',                                    version:  7, name: 10
        name: 
        capabilities: pointer keyboard touch
        keyboard repeat rate: 25
        keyboard repeat delay: 660
interface: 'zwp_pointer_gestures_v1',                    version:  3, name: 11
interface: 'zwp_pointer_constraints_v1',                 version:  1, name: 12
interface: 'zwp_relative_pointer_manager_v1',            version:  1, name: 13
interface: 'wl_data_device_manager',                     version:  3, name: 14
interface: 'zwlr_data_control_manager_v1',               version:  2, name: 15
interface: 'zwp_primary_selection_device_manager_v1',    version:  1, name: 16
interface: 'org_kde_kwin_idle',                          version:  1, name: 17
interface: 'zwp_idle_inhibit_manager_v1',                version:  1, name: 18
interface: 'org_kde_plasma_shell',                       version:  8, name: 19
interface: 'org_kde_kwin_appmenu_manager',               version:  1, name: 20
interface: 'org_kde_kwin_server_decoration_palette_manager', version:  1, name: 21
interface: 'org_kde_plasma_virtual_desktop_management',  version:  2, name: 23
interface: 'org_kde_kwin_shadow_manager',                version:  2, name: 25
interface: 'org_kde_kwin_dpms_manager',                  version:  1, name: 26
interface: 'org_kde_kwin_server_decoration_manager',     version:  1, name: 27
interface: 'kde_output_management_v2',                   version:  2, name: 28
interface: 'kde_primary_output_v1',                      version:  2, name: 29
interface: 'zxdg_output_manager_v1',                     version:  3, name: 30
        xdg_output_v1
                output: 47
                name: 'eDP-1'
                description: 'Sharp Corporation eDP-1-unknown'
                logical_x: 0, logical_y: 0
                logical_width: 1422, logical_height: 800
interface: 'wl_subcompositor',                           version:  1, name: 31
interface: 'zxdg_exporter_v2',                           version:  1, name: 32
interface: 'zxdg_importer_v2',                           version:  1, name: 33
interface: 'xdg_activation_v1',                          version:  1, name: 35
interface: 'wp_drm_lease_device_v1',                     version:  1, name: 37
interface: 'wl_drm',                                     version:  2, name: 43
interface: 'zwp_linux_dmabuf_v1',                        version:  4, name: 44
        formats:
        'UYVY'(0x59565955), modifier: 0x0100000000000002
        'UYVY'(0x59565955), modifier: 0x0100000000000001
        'UYVY'(0x59565955), modifier: 0x0000000000000000
        'R16 '(0x20363152), modifier: 0x0100000000000002
        'R16 '(0x20363152), modifier: 0x0100000000000001
        'R16 '(0x20363152), modifier: 0x0000000000000000
        'GR32'(0x32335247), modifier: 0x0100000000000004
        'GR32'(0x32335247), modifier: 0x0100000000000002
        'GR32'(0x32335247), modifier: 0x0100000000000001
        'GR32'(0x32335247), modifier: 0x0000000000000000
        'XB48'(0x38344258), modifier: 0x0100000000000004
        'XB48'(0x38344258), modifier: 0x0100000000000002
        'XB48'(0x38344258), modifier: 0x0100000000000001
        'XB48'(0x38344258), modifier: 0x0000000000000000
        'XR24'(0x34325258), modifier: 0x0100000000000004
        'XR24'(0x34325258), modifier: 0x0100000000000002
        'XR24'(0x34325258), modifier: 0x0100000000000001
        'XR24'(0x34325258), modifier: 0x0000000000000000
        'AB24'(0x34324241), modifier: 0x0100000000000004
        'AB24'(0x34324241), modifier: 0x0100000000000002
        'AB24'(0x34324241), modifier: 0x0100000000000001
        'AB24'(0x34324241), modifier: 0x0000000000000000
        'RG16'(0x36314752), modifier: 0x0100000000000002
        'RG16'(0x36314752), modifier: 0x0100000000000001
        'RG16'(0x36314752), modifier: 0x0000000000000000
        'XR30'(0x30335258), modifier: 0x0100000000000002
        'XR30'(0x30335258), modifier: 0x0100000000000001
        'XR30'(0x30335258), modifier: 0x0000000000000000
        'P010'(0x30313050), modifier: 0x0100000000000002
        'P010'(0x30313050), modifier: 0x0100000000000001
        'P010'(0x30313050), modifier: 0x0000000000000000
        'Y412'(0x32313459), modifier: 0x0100000000000002
        'Y412'(0x32313459), modifier: 0x0100000000000001
        'Y412'(0x32313459), modifier: 0x0000000000000000
        'AB30'(0x30334241), modifier: 0x0100000000000002
        'AB30'(0x30334241), modifier: 0x0100000000000001
        'AB30'(0x30334241), modifier: 0x0000000000000000
        'XB4H'(0x48344258), modifier: 0x0100000000000004
        'XB4H'(0x48344258), modifier: 0x0100000000000002
        'XB4H'(0x48344258), modifier: 0x0100000000000001
        'XB4H'(0x48344258), modifier: 0x0000000000000000
        'Y210'(0x30313259), modifier: 0x0100000000000002
        'Y210'(0x30313259), modifier: 0x0100000000000001
        'Y210'(0x30313259), modifier: 0x0000000000000000
        'Y416'(0x36313459), modifier: 0x0100000000000002
        'Y416'(0x36313459), modifier: 0x0100000000000001
        'Y416'(0x36313459), modifier: 0x0000000000000000
        'YUYV'(0x56595559), modifier: 0x0100000000000002
        'YUYV'(0x56595559), modifier: 0x0100000000000001
        'YUYV'(0x56595559), modifier: 0x0000000000000000
        'AB48'(0x38344241), modifier: 0x0100000000000004
        'AB48'(0x38344241), modifier: 0x0100000000000002
        'AB48'(0x38344241), modifier: 0x0100000000000001
        'AB48'(0x38344241), modifier: 0x0000000000000000
        'AR24'(0x34325241), modifier: 0x0100000000000004
        'AR24'(0x34325241), modifier: 0x0100000000000002
        'AR24'(0x34325241), modifier: 0x0100000000000001
        'AR24'(0x34325241), modifier: 0x0000000000000000
        'AR15'(0x35315241), modifier: 0x0100000000000002
        'AR15'(0x35315241), modifier: 0x0100000000000001
        'AR15'(0x35315241), modifier: 0x0000000000000000
        'XYUV'(0x56555958), modifier: 0x0100000000000002
        'XYUV'(0x56555958), modifier: 0x0100000000000001
        'XYUV'(0x56555958), modifier: 0x0000000000000000
        'GR88'(0x38385247), modifier: 0x0100000000000002
        'GR88'(0x38385247), modifier: 0x0100000000000001
        'GR88'(0x38385247), modifier: 0x0000000000000000
        'AR30'(0x30335241), modifier: 0x0100000000000002
        'AR30'(0x30335241), modifier: 0x0100000000000001
        'AR30'(0x30335241), modifier: 0x0000000000000000
        'R8  '(0x20203852), modifier: 0x0100000000000002
        'R8  '(0x20203852), modifier: 0x0100000000000001
        'R8  '(0x20203852), modifier: 0x0000000000000000
        'AB4H'(0x48344241), modifier: 0x0100000000000004
        'AB4H'(0x48344241), modifier: 0x0100000000000002
        'AB4H'(0x48344241), modifier: 0x0100000000000001
        'AB4H'(0x48344241), modifier: 0x0000000000000000
        'P012'(0x32313050), modifier: 0x0100000000000002
        'P012'(0x32313050), modifier: 0x0100000000000001
        'P012'(0x32313050), modifier: 0x0000000000000000
        'XB24'(0x34324258), modifier: 0x0100000000000004
        'XB24'(0x34324258), modifier: 0x0100000000000002
        'XB24'(0x34324258), modifier: 0x0100000000000001
        'XB24'(0x34324258), modifier: 0x0000000000000000
        'P016'(0x36313050), modifier: 0x0100000000000002
        'P016'(0x36313050), modifier: 0x0100000000000001
        'P016'(0x36313050), modifier: 0x0000000000000000
        'Y212'(0x32313259), modifier: 0x0100000000000002
        'Y212'(0x32313259), modifier: 0x0100000000000001
        'Y212'(0x32313259), modifier: 0x0000000000000000
        'Y410'(0x30313459), modifier: 0x0100000000000002
        'Y410'(0x30313459), modifier: 0x0100000000000001
        'Y410'(0x30313459), modifier: 0x0000000000000000
        'AYUV'(0x56555941), modifier: 0x0100000000000002
        'AYUV'(0x56555941), modifier: 0x0100000000000001
        'AYUV'(0x56555941), modifier: 0x0000000000000000
        'Y216'(0x36313259), modifier: 0x0100000000000002
        'Y216'(0x36313259), modifier: 0x0100000000000001
        'Y216'(0x36313259), modifier: 0x0000000000000000
interface: 'kde_output_device_v2',                       version:  2, name: 46
interface: 'wl_output',                                  version:  4, name: 47
        x: 0, y: 0, scale: 3,
        physical_width: 290 mm, physical_height: 170 mm,
        make: 'Sharp Corporation', model: 'eDP-1-unknown',
        subpixel_orientation: unknown, output_transform: normal,
        mode:
                width: 3200 px, height: 1800 px, refresh: 59.982 Hz,
                flags: current
interface: 'zwp_text_input_manager_v2',                  version:  1, name: 48
interface: 'zwp_text_input_manager_v3',                  version:  1, name: 49
interface: 'org_kde_kwin_contrast_manager',              version:  2, name: 51
interface: 'org_kde_kwin_blur_manager',                  version:  1, name: 52
interface: 'org_kde_kwin_slide_manager',                 version:  1, name: 53
```

edwintorok commented:
Thanks, looks like 'wl_output, scale' would be correct, or not too far off (3 for you, 2 for me), and we could also calculate DPI from physical width/height vs pixel width/height.
Looking at xdpyinfo its output is completely bogus under wayland:
```
screen #0:
  dimensions:    7680x2160 pixels (2032x571 millimeters)
  resolution:    96x96 dots per inch
```

However 'xrandr' seems OKish:
```
xrandr
Screen 0: minimum 16 x 16, current 7680 x 2160, maximum 32767 x 32767
XWAYLAND0 connected primary 3840x2160+3840+0 (normal left inverted right x axis y axis) 530mm x 300mm
```

Could you also paste your 'xrandr' output please? (it is more likely that xrandr is preinstalled than wayland-info)

sanette commented:
`xdpyinfo` is indeed bogus:

```
screen #0:
  dimensions:    3200x1800 pixels (846x476 millimeters)
  resolution:    96x96 dots per inch
```

`xrandr` is good:

```
$ xrandr 
Screen 0: minimum 16 x 16, current 3200 x 1800, maximum 32767 x 32767
XWAYLAND0 connected primary 3200x1800+0+0 (normal left inverted right x axis y axis) 290mm x 170mm
   3200x1800     59.96*+
```

sanette commented:
Note also that, at least for me, under Wayland SDL chooses the x11 driver as default, and then everything works well.
See also https://www.phoronix.com/news/SDL2-Reverts-Wayland-Default
